File: designPatterns/SOLID_design_principles/InterfaceSegregationPrinciple/typescript/violateInterfaceSegregationPrinciple.ts
// this violates the interface segregation principle
// because it is being used by multiple classes that does not requires all of it
interface Entity {
  name: string;
  attackDamage: number;
  health: number;

  move(): void | null;
  attack(targetEntity: Entity | undefined): void | null;
  takeDamage(amount: number): void | null;
}

// passes the interface segregation principle
class Character implements Entity {
  name: string;
  attackDamage: number;
  health: number;

  constructor(name: string, attackDamage: number, health: number) {
    this.name = name;
    this.attackDamage = attackDamage;
    this.health = health;
  }

  move(): void {
    console.log(`${this.name} moved`);
  }

  attack(targetEntity: Entity | undefined): void {
    console.log(
      `${this.name} attacked ${targetEntity.name} for ${this.attackDamage} damage`,
    );
    targetEntity.takeDamage(this.attackDamage);
  }

  takeDamage(amount: number | undefined): void {
    this.health -= amount;
    console.log(`${this.name} has ${this.health} health remaining`);
  }
}

class Wall implements Entity {
  name: string;
  // violates the interface segregation principle
  attackDamage: number;
  health: number;

  constructor(name: string, health: number) {
    this.name = name;
    this.attackDamage = 0;
    this.health = health;
  }

  // violates the interface segregation principle
  move(): null {
    return null;
  }

  // violates the interface segregation principle
  attack(): null {
    return null;
  }

  takeDamage(amount: number): void {
    this.health -= amount;
    console.log(`${this.name} has ${this.health} health remaining`);
  }
}

class Turret implements Entity {
  name: string;
  attackDamage: number;
  // violates the interface segregation principle
  health: number;

  constructor(name: string, attackDamage: number) {
    this.name = name;
    this.attackDamage = attackDamage;
    this.health = 0;
  }

  // violates the interface segregation principle
  move(): null {
    return null;
  }

  attack(targetEntity: Entity): void {
    console.log(
      `${this.name} attacked ${targetEntity.name} for ${this.attackDamage} damage`,
    );
    targetEntity.takeDamage(this.attackDamage);
  }

  // violates the interface segregation principle
  takeDamage(): null {
    return null;
  }
}

const turret = new Turret("turret", 5);
const character = new Character("Character", 3, 100);
const wall = new Wall("Wall", 200);

turret.attack(character);
character.move();
character.attack(wall);

File: designPatterns/SOLID_design_principles/InterfaceSegregationPrinciple/python/violateInterfaceSegregationPrinciple.py
from abc import ABC, abstractmethod

# this violates the interface segregation principle
# because it is being used by multiple classes that does not requires all of it
class Entity(ABC):
    def __init__(self, name, attack_damage, health):
        self.name = name
        self.attack_damage = attack_damage
        self.health = health

    @abstractmethod
    def move(self):
        pass

    @abstractmethod
    def attack(self, target_entity):
        pass

    @abstractmethod
    def take_damage(self, amount):
        pass


class Character(Entity):
    def __init__(self, name, attack_damage, health):
        super().__init__(name, attack_damage, health)

    def move(self):
        print(f"{self.name} moved")

    def attack(self, target_entity):
        print(
            f"{self.name} attacked {target_entity.name} for {self.attack_damage} damage",
        )
        target_entity.take_damage(self.attack_damage)

    def take_damage(self, amount):
        self.health -= amount
        print(f"{self.name} has {self.health} health remaining")


class Wall(Entity):
    def __init__(self, name, health):
        super().__init__(name, 0, health)

    def move(self):
        return None

    def attack(self):
        return None

    def take_damage(self, amount):
        self.health -= amount
        print(f"{self.name} has {self.health} health remaining")


class Turret(Entity):
    def __init__(self, name, attack_damage):
        super().__init__(name, attack_damage, 0)

    def move(self):
        return None

    def take_damage(self):
        return None

    def attack(self, target_entity):
        print(
            f"{self.name} attacked {target_entity.name} for {self.attack_damage} damage",
        )
        target_entity.take_damage(self.attack_damage)


turret = Turret("Turret", 5)
character = Character("Character", 3, 100)
wall = Wall("Wall", 200)

turret.attack(character)
character.move()
character.attack(wall)
